hal/inc/socket_hal.h

```diff
@@ -92,7 +92,21 @@ sock_result_t socket_accept(sock_handle_t sd);
  */
 sock_handle_t socket_handle_invalid();
 
+/*
+ * Join a multicast address for all UDP sockets. This will allow reception of multicast packets
+ * sent to the given address on all UDP sockets which have bound the port to which the multicast
+ * packet was sent.
+ * @param addr IP multicast address to join
+ * @return Return the result of the join operation
+ */
+sock_result_t socket_join_multicast(const sockaddr_t *addr);
```

Comment:
Please parameterize with a `network_interface_t` value since the Photon will soon support multiple networks.  I wonder if this is a good reason to make it an instance method on UDP rather than static, so we know the network interface to use?

Finally we need one or more void* reserved param at the end for future expansion. (Once set, the function signatures cannot change.)